### ISSUE TITLE
methodop ». / methodop >>.

### ISSUE THREAD

**Issue by 0rir** [Raku/doc#4108]
In `language/operator`, in the precedence list, the line

> Method call	left	.meth .+ .? .* .() .[] .{} .<> .«» .:: .= .^ .: i

lacks the `».` operator.

**Comment by 2colours** [comment]
I'm still wondering whether this could be intentional. If you think about it, `»` is a meta-operator. Meta-operators usually don't appear in the precedence list.

**Comment by raiph** [comment]
@0rir

> In language/operator, in the precedence list, the line
>
> | Precedence Level | Associativity | Examples |
> | -- | -- | -- |
> | Method call | left | `.meth` `.+` `.?` `.*` `.()` `.[]` `.{}` `.<>` `.«»` `.::` `.=` `.^` `.:` `i` |
>
> lacks the `».` operator.

As a perhaps minor point, I see that the title of the third column of the table is "Examples". That suggests to me a deliberate decision not to promise an exhaustive list. (It may be that the table originally *did* promise an exhaustive list and then someone altered the title of the third column to "Examples". I don't know, but it's definitely "Examples" in the rendering I'm looking at.) So from at least that perspective it's presumably perfectly cromulent not to include `».`.

----

It's perhaps interesting to note discrepancies between this table and the other precedence list on the same page, the one on the left of the page. (I tend to refer to that list instead of the table because A) it appears to be more exhaustive in intent, and B) it lists operators in order of precedence, unlike the table. cf #4085 et al.)

The list of ops on the left includes a corresponding "Method postfix" section. Here's the ops it lists, expressed in the same form and ordering as the line in the table above:

> Method postfix precedence
> 
> `.` `.&` `.+` `.?` `.*`  `.()` `.[]` `{}` `.<>` `.< >` `.«»` `.::` `.=` `.^` `.:<prefix operator>` `».` / `>>.` `.postfix` / `.postcircumfix` `,=`

Here are entries in the (exhaustive?) list's "Method postfix" section but not the table's corresponding "Examples" row:

`.&` `».` / `>>.` `.postfix` / `.postcircumfix` `,=`

(I didn't include `.` because I take it to correspond to the table's `.meth`. Likewise `.:<prefix operator>` because I take it to correspond to the table's `.:`.)

So the table omits 5 entries that are in the list. **One of these is is `».`.**

(I'm not sure why `,=` is in the list. `,=` is just an `op=` compound op at first blush, but I vaguely recall there being something special about it; perhaps there's some good reason to explicitly list it.)

Going in the other direction, here are the "Examples" in the table's "Method call" row that aren't in the (exhaustive?) list's corresponding section:

 `i`

So, just the one. (And as examples go, `i` seems like an... interesting choice to include. I shan't comment on it further but didn't want to omit at least a mention of it.

**Comment by 2colours** [comment]
> B) (I thought I thought, but as I write this wonder if I'm losing my mind) it lists operators in order of precedence, unlike the table. cf https://github.com/Raku/doc/issues/4071 et al.)

I think you are right. Something got broken around a fixup of the operator list but the links are in a somewhat better shape. Also, alas, we basically cannot deploy to that site at the moment and everyone is hoping for a swift resolution for the [new docs site](https://github.com/Altai-man/docs.raku.org).

> (I'm not sure why ,= is in the list. ,= is just an op= compound op at first blush, but I vaguely recall there being something special about it; perhaps there's some good reason to explicitly list it?)

Funny that you point this one out; it's still kinda foggy if `=` is considered a meta-operator in Raku or Rakudo, and one thing is sure - it's causing a lot of confusion with `.=` (which has "methodop precedence" documentedly, but... why).

**Comment by 0rir** [comment]
@raiph, I accept that your point that the table is of 'examples' invalidates this issue.  That is rather unsatisfying since the  page's ToC is not doc'ed to be a precedence table.  And that there are errors within a single line of the examples table.

 I am thinking of closing this and raising an issue that `a separate page be created as a traditional precedence table`.  I am sad to stack up work that I cannot do;  especially since I generally find  Raku's precedence structure to _just work_ .

@2colours, I don't consider `=` a meta op; and I call `WHATEVER=` compound ops.  While `.WHATEVER` are deviant method ops.

**Comment by 2colours** [comment]
@0rir what do you think about this issue? Can it be closed, or if not, how to make it actionable?

**Comment by 0rir** [comment]
@2colours, I don't see how it is not actionable; other than by lack of agreement, interest or time.

I decided against dumping this and raising the idea of a separate precedence page, because there are then three datasets to resolve:  the 'partial precedence table',  the operator page's ToC, and the new page.

A precedence table could  be appended to the end of the operator page.

I am patient.  And especially so, given that the AST has been suggested as a good tool for getting alternate actions from the Raku grammar:  linting, cataloging usage, maybe even this issue.

**Comment by 2colours** [comment]
> @2colours, I don't see how it is not actionable; other than by lack of agreement, interest or time.

I mean, it is not actionable literally. It doesn't say what the desired outcome is and lies on outdated presumptions that I at least thought were well-understood.

1. the "precedence table" is ordered by precedence again, that issue has been resolved in the meantime.
2. it is actually known to be non-exhaustive - it even says "examples", and you didn't explicitly object to that?
3. it's not clear at all what "exhaustive" would mean. Should all `op=` operators be listed? Is there even a point when `op=` is a productive formation, i.e if you define `lol` as an operator, `lol=` will just work? (In this sense, `=` _is surely_ a meta-operator.)
4. It might make sense to add `».` to the list. _That_ is actionable, and doesn't address the "exhaustiveness issue".  
Even though, I think that begs the question what makes it stand out of usual hyper meta-operator stuff, as it seems to obey the same syntax rules. From a "precedence table" point of view, the dot is not distinguished, even though it isn't a first-class operator, just syntax in the parser. Consequently, no meta-operator can work with it "naively", and the result will always be another piece of syntax, rather than first-class citizens of the language. That is worth marking but it should be done in a consistent manner...

**Comment by 0rir** [comment]
@2colours and  @raiph,  you have made good points that make my issue trivial as titled.  By that closing the issue makes sense.

So this conversation is misplaced and the blame is mine.   But with a view toward forming a better issue, to @2colours'  points:

1. That has improved the situation greatly.
2. I accepted (then and now that the "Examples" title  invalidates this issue.   But if one is looking for a precedence they may ignore the preamble entirely and through practice accept it as the precedence table.
3.  "Exhaustive" ~~ "Complete".   My take:   For some reasonably inclusive definition of potential  racruits it should make the various operators recognizable and their relative precedence levels understandable.  (Using my personal terminology previously was unhelpful.)
4. Adding `».` does address the  "exhaustiveness"  in a small way.   `».` _might_ stand out as one of the commonest of dot constructions.      The doc's `Technically, not a real operator; it's syntax special-cased in the compiler.` is pretty useless to one reading the page to use ops at a basic level.    Ops are doc'ed on three separate pages now and those are not well connected (when last I checked).
